src/storage/src/hummock/local_version_manager.rs

```diff
@@ -220,16 +220,23 @@ impl LocalVersionManager {
                 return false;
             }
         }
-        let mut guard = self.local_version.write();
 
-        if guard.pinned_version().id() >= new_version_id {
+        let old_version = self.local_version.read().clone();
+        if old_version.pinned_version().id() >= new_version_id {
             return false;
         }
 
         if let Some(conflict_detector) = self.write_conflict_detector.as_ref() {
             conflict_detector.set_watermark(newly_pinned_version.max_committed_epoch);
         }
-        guard.set_pinned_version(newly_pinned_version);
+
+        let mut new_version = old_version;
+        new_version.set_pinned_version(newly_pinned_version);
+        {
+            let mut guard = self.local_version.write();
+            *guard = new_version;
```

Comment:
I'm not sure how is the difference between `move one large object` and `move one Arc pointer`.